### DISCUSSION TITLE
LPC failsafe write fails

### DISCUSSION THREAD

**Discussion by heavyweight87** [enbility/eebus-go#114] [Q&A] [answered]
Hi,

When writting the failsafe duration or limit as an EG I get the error on the CS side: 

**_update failed, likely not allowed to write_**

I cant figure out why this happens - any idea? 

I am using WriteFailsafeConsumptionActivePowerLimit and WriteFailsafeDurationMinimum

BTW errors are sent to the debug log.... maybe its worth adding an error log type as well? 

Thanks!

**Comment by DerAndereAndi** [comment]
Did you initialize the CS data storage using the available public methods?

These kind of errors are send on debug on purpose for now.

I would need an actual log and code example to say what is happening. The above mentioned is the most likely cause. Also the version of the libraries are relevant.

**Comment by heavyweight87** [comment]
Its with the latest version

Init storage how? Sounds like this is probably it...
Could it be that I should call SetFailsafeConsumptionActivePowerLimit for example, on the CS side? 

Setting the consumption limit works if that helps

**Comment by DerAndereAndi** [comment]
I assume latest version is the latest release. There the default is to not allow write.

You can either use the latest commit on the dev branch, which changed the default, or you can call `SetFailsafeConsumptionActivePowerLimit` to set `changeable` to true.

In general you should initialize all data structures in a controllable system to your desired default values. The stack can't know these. And always do this before `h.myService.Start()` is invoked.

**Reply by heavyweight87** [reply]
Yes I meant release. 

 I have something like the following, but it didn't seem to help


``` go
// Adds the use cases to the service.
func (cems *Cems) addUseCases() {
	localEntity := cems.service.LocalDevice().EntityForType(model.EntityTypeTypeCEM)
	cems.uccslpc = cslpc.NewLPC(localEntity, cems.OnLPCEvent)
	cems.service.AddUseCase(cems.uccslpc)
	cems.uccslpp = cslpp.NewLPP(localEntity, cems.OnLPPEvent)
	cems.service.AddUseCase(cems.uccslpp)
	cems.uceglpc = eglpc.NewLPC(localEntity, cems.OnLPCEvent)
	cems.service.AddUseCase(cems.uceglpc)
	cems.uceglpp = eglpp.NewLPP(localEntity, cems.OnLPPEvent)
	cems.service.AddUseCase(cems.uceglpp)
	cems.ma = ucma.NewMPC(localEntity, nil)
	cems.service.AddUseCase(cems.ma)

	//set defaults
	cems.uccslpc.SetFailsafeConsumptionActivePowerLimit(1000, true)
	cems.uccslpc.SetFailsafeDurationMinimum(10*time.Second, true)
}
```


This is done before service.start() obviously

**Reply by DerAndereAndi** [reply, marked as answer]
Please try the dev branch. In my test it works there

**Reply by heavyweight87** [reply]
Thanks it works great now!

**Comment by msaleme** [comment]
Great thread. A couple of things worth flagging based on what we have been testing:

@DrCookies84 API keys bound to agent IDs is solid for the single-hop case. Where it gets interesting is multi-hop delegation - Agent A calls Agent B with valid credentials, but B has no way to verify whether A's human approved *that specific delegation*. The credential says "A is authorized" but not "A is authorized to ask B to do this particular thing." That is the gap AG-003 and the new delegation chain tests are designed to surface.

@jee599 Cryptographic signing per agent identity is the right call. The hard part is adoption. Most frameworks treat it as optional hardening rather than default behavior. We have been experimenting with lightweight HMAC chains on the delegation path - enough to verify provenance without adding latency that makes developers rip it out.

Re: pointing the harness at NULL's API - happy to do that. Would be a good mutual stress test. Open an issue on the repo or DM and we can coordinate.
